keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,228 @@
+
+Components should plumb the context through reconciliation functions, rather than storing and looking up trace contexts globally so that each attempt to reconcile desired and actual state uses the context associated with _that_ desired state through the entire attempt.  If multiple components are involved in reconciling a single object, one may act on the new trace context before the other, but each trace is still representative of the work done to reconcile to the corresponding desired state. Given this model, we guarantee that each trace contains the actions taken to reconcile toward a single desired state.
+
+High-level processes, such as starting a pod or restarting a failed container, could be interrupted before completion by an update to the desired state. While this leaves a "partial" trace for the first process, it is the most accurate representation of the work and timing of reconciling desired and actual state.
+
+#### End-User Behavior
+
+A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+
+Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+
+For example, a user could define a Pod Creation trace doing something like the following:
+```golang
+ctx, span := trace.StartSpan(context.Background(), “create-my-pod”)
+defer span.End()
+ctx = traceutil.WithSpanContext(ctx, span)
+pod, err := c.CoreV1().Pod(myPod.Namespace).Create(ctx, myPod)
+if err != nil {
+    return err
+}
+waitForPodToBeRunning(ctx, myPod)
+return nil
+```
+
+### In-tree changes
+
+#### Vendor the Tracing Framework
+
+This KEP proposes the use of the [OpenTelemetry tracing framework](https://opentelemetry.io/) to create and export spans to configured backends. However, this framework is still under development.  During the alpha phase, we will use [OpenCensus](https://opencensus.io/), which promises to be compatible with opentelemetry when it is released. 
```

Comment:
How was this picked over opentracing?

Reply:
IIRC (its been a year since we made that decision...), the biggest factor was being able to do exporter configuration out-of-tree in the agent.  I tend to value that pretty highly given the k8s communities poor experiences with in-tree storage exporters with cAdvisor + heapster... I'm happy to reconsider if there are other reasons I haven't considered.

Reply:
I don't have enough experience with either to have an opinion, just wondered how the decision was made. It looks like opentracing doesn't ship with exporters so opencensus makes sense if we are going to use the agent.

Reply:
AFAIK OpenTracing was only ever an API spec (with a light reference implementation). OpenTelemetry is an effort to unify OpenCensus (which provided impl for various languages / exporters to a bunch of backends) and OpenTracing in the sense that the shipped tracing libraries will conform to what was previously the OpenTracing API spec ([from here](https://medium.com/opentracing/a-roadmap-to-convergence-b074e5815289)). 

OpenCensus and OpenTracing will both be sunset in November, so OpenTelemetry should be the only library mentioned here.